### ISSUE TITLE
Release v0.23.1 (please TEST!)

### ISSUE THREAD

**Issue by Stypox** [TeamNewPipe/NewPipe#8548]
## Testing for regressions

Debug APK (built by our CI in #8547): [NewPipe_v0.23.1_RC1_debug.zip](https://github.com/TeamNewPipe/NewPipe/files/8980036/NewPipe_v0.23.1_RC1_debug.zip)
Please report **only regressions** (i.e. new issues) here, not issues that were already present in the previous release!

## TODO

- [ ] Merge weblate
- [ ] Release a NewPipeExtractor version
- [ ] #8531
- [ ] #8336 needs to be reverted or really fixed

## NewPipeExtractor version

This version of NewPipe will ship with [NewPipeExtractor version 0.23.0](https://github.com/TeamNewPipe/NewPipeExtractor/releases/tag/v0.23.0)
-> the NPE release still hasn't been done ;-)

### New
- Support delivery methods other than progressive HTTP #8153 (thanks to @AudricV!)
  - [YouTube] Use DASH for playback, improving loading times in playback and seeking
  - [YouTube] Full playback of recently ended livestreams
  - [YouTube] Improve availability of some resolutions and formats
  - [PeerTube] Fix crash on videos with an audio-only stream
  - [PeerTube] Fix seeking on videos with only HLS streams
  - [SoundCloud] Fix playback of SoundCloud HLS-only tracks

- Add button to add a remote playlist to a local one #7355
- Image preview in Android share sheet (Android 10+ only) #8143 

### Improved
- Improve playback parameters dialog's UI #7989 #8349
- Moved subscription import/export options to (overflow) menu #7458

### Fixed
- Fix keyboard doesn't open with disabled animations #8227
- Fix crash when open NewPipe from notification bar #8244
- Fix hiding player controls when playing from media button #8315
- Fix viewed counting in history #8336 (**TODO: ⚠️ needs to be reverted or really fixed**)
- Fix `RouterActivity` theming *(the menu that shows up when you share something to NewPipe)* #8332
- Fix removing only fully watched videos from playlist #8259
- Fix crash when rotating device on unsupported channels #8192
- Make volume progress bar match system volume when we start sliding #8377
- Fix `Add to playlist` not working and cleanup `RouterActivity` choice handling *(the menu that shows up when you share something to NewPipe)* #8340

### Localization
- Add Hindi translation of the README #8355

### Development
- Update ACRA library #8206
- Update user agent used by the downloader #8269
- Remove unused `strings.xml` resources #8329
- Update Groupie and ACRA libraries #8407
- Update Android Gradle Plugin and Kotlin #8406
- Use JvmOverloads instead of manually overloading method #8391
- Remove unnecessary compat method calls #8382
- Use `AppCompatResources` #8404

**Comment by Nizami20052022** [comment]
I have added Azerbaijani language translations to be updated.

**Comment by swarup13** [comment]
Please release this version.

**Comment by Nizami20052022** [comment]
Write this comment in the Pull requests section.

**Comment by avently** [comment]
Network operations like viewing youtube videos and seeking is so much faster than before, so cool

**Comment by opusforlife2** [comment]
Livestreams often error out due to source errors, but I haven't seen this HLS one yet. Not sure if it's a regression or just another regular livestream error.

## Exception
* __User Action:__ play stream
* __Request:__ Player error[type=ERROR_CODE_IO_UNSPECIFIED] occurred while playing https://www.youtube.com/watch?v=gvpE6615qec
* __Content Country:__ CH
* __Content Language:__ en-IN
* __App Language:__ en_IN
* __Service:__ YouTube
* __Version:__ 0.23.1
* __OS:__ Linux Android 10 - 29
<details><summary><b>Crash log </b></summary><p>

```
com.google.android.exoplayer2.ExoPlaybackException: Source error
	at com.google.android.exoplayer2.ExoPlayerImplInternal.handleIoException(ExoPlayerImplInternal.java:641)
	at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:617)
	at android.os.Handler.dispatchMessage(Handler.java:103)
	at android.os.Looper.loop(Looper.java:214)
	at android.os.HandlerThread.run(HandlerThread.java:67)
Caused by: com.google.android.exoplayer2.source.hls.playlist.HlsPlaylistTracker$PlaylistStuckException
	at com.google.android.exoplayer2.source.hls.playlist.DefaultHlsPlaylistTracker$MediaPlaylistBundle.processLoadedPlaylist(DefaultHlsPlaylistTracker.java:739)
	at com.google.android.exoplayer2.source.hls.playlist.DefaultHlsPlaylistTracker$MediaPlaylistBundle.onLoadCompleted(DefaultHlsPlaylistTracker.java:584)
	at com.google.android.exoplayer2.source.hls.playlist.DefaultHlsPlaylistTracker$MediaPlaylistBundle.onLoadCompleted(DefaultHlsPlaylistTracker.java:512)
	at com.google.android.exoplayer2.upstream.Loader$LoadTask.handleMessage(Loader.java:478)
	at android.os.Handler.dispatchMessage(Handler.java:107)
	... 2 more

```
</details>
<hr>

**Comment by opusforlife2** [comment]
Oh, and I don't know if this is a regression. YT Shorts thumbnails are now fit to width, cutting off the top and bottom portions, instead of fit to height, with black bars on the side, like in the current release.

**Comment by AudricV** [comment]
Ended age-restricted livestreams cannot be watched anymore.

Reason: the client used to bypass age-restricted contents only provides a DASH and a HLS manifest, which are incomplete, and no audio or video stream seems to be provided (issue was reproduced on video `PXvs97_7z2I`, now rencoded as a video and non-age restricted) -> a `StreamExtractException` exception with the message `Could not get any stream` is thrown.

In order to fix this issue, we have to parse the DASH manifest to extract `Stream`s objects and:

- calculate the average duration between two segments;
- extract the base streaming URL;
- build an `ItagItem`;
- make `YoutubePostLiveStreamDvrDashManifestCreator` accepts "parameters and values" using `/` instead of `&` and `=`;
- and probably more.

Screenshot of the `streamingData` object returned in this case for the client used to bypass age-restricted contents:

<img src="https://user-images.githubusercontent.com/74829229/176474876-65596fa7-1ee4-4d3e-8aa4-d78255069066.png" width="500px" alt="YouTube streamingData postLiveDvr age-restricted contents">

**Comment by AudricV** [comment]
@opusforlife2

> Livestreams often error out due to source errors, but I haven't seen this HLS one yet. Not sure if it's a regression or just another regular livestream error.

That's the exception (`PlaylistStuckException`) for which coefficient was slightly increased. There is nothing we can do about this, and is related to a server-side issue (the error means that the manifest has been not updated since a while, if I am not wrong). It can also happen on the end of brodcasts, because YouTube doesn't end properly HLS manifests sometimes.

> This one isn't a livestream, but a normal video. And I haven't seen this error before either.

Were you listening to the content or were you watching it? The `TimeoutException`s may happen in background because of battery optimizations, especially the agressive ones of OEMs, which can do sometimes strange behaviors on networks state. I got this issue probably at most three times row since I use #8153 (for two months at least and maybe more) on my phone running EMUI. If you were watching the content, then that's probably a network/server issue.

We can also increase the connection timeout, if that's needed.

> Oh, and I don't know if this is a regression. YT Shorts thumbnails are now fit to width, cutting off the top and bottom portions, instead of fit to height, with black bars on the side, like in the current release.

That's a YouTube change on short thumbnails (which are returned without "borders" (so not anymore in the 16/9 format) on which we can't really do something, except that we can remove the `sqp` and/or the `rs` parameters to get original thumbnails from the image URL to get again a 16/9 format.

The only issue is that thumbnails which have normally black bars but don't have them using these parameters will get black bars, which isn't something we'd like to have I think.

Also, shorts thumbnail URLs seem to use a `_2` in the end of the  thumbnail name.

Finally, contrary to videos, there is now only one thumbnail returned for shorts:

<img src="https://user-images.githubusercontent.com/74829229/176962136-5f4bdb3c-fb0c-41f5-b71f-760e14ba2cfe.png" width="400px" alt="InnerTube_trending_shorts_thumbnails">

In conclusion, this change is not a regression at all (and is also reproducible on the current release for me).

**Comment by opusforlife2** [comment]
> If you were watching the content, then that's probably a network/server issue. We can also increase the connection timeout, if that's needed.

I was watching. I was thinking that there might be need somewhere in the code to request YT to refresh the connection, instead of just increasing the timeout? If timeout is the only option, then yeah, increasing that might be good. Although I think we might need to test if that helps at all.

**Comment by opusforlife2** [comment]
> The only issue is that thumbnails which have normally black bars but don't have them using these parameters will get black bars, which isn't something we'd like to have I think.

Why? If it's just Shorts thumbnails displaying this behaviour, they will always be in 16:9 aspect ratio, and should have black bars.

Are you saying the proposed code change will affect non-Shorts thumbnails as well, and there is no way to prevent that?

**Comment by Donkey-Doug** [comment]
FYI, this completely fixes the intermittent, constant buffering issue for me.

**Comment by Stypox** [comment]
@opusforlife2 according to what @AudricV explained above, thumbnails are now extracted at their original size, instead of at the forced-16:9 size like before (which included black bars for shorts). So the extractor has only improved and there is no regression there. The only problem is that thumbnails in NewPipe are not rendered in an optimal way, cropping them instead of showing them whole. I'll open a PR about this.

**Comment by Stypox** [comment]
I would like to release tomorrow, if nothing blocking comes up. I updated the apk with the latest (small) changes, and also provided a release APK if someone wants to test it out. :-)
